src/Commands/Remote/SSHBaseCommand.php

```diff
@@ -70,28 +70,80 @@ protected function setProgressAllowed($allowed = true)
      */
-    protected function executeCommand(array $command_args)
+    protected function executeCommand(array $command_args, int $retries = 0)
     {
         $this->validateEnvironment($this->environment);
 
         $command_summary = $this->getCommandSummary($command_args);
         $command_line = $this->getCommandLine($command_args);
 
-        $ssh_data = $this->sendCommandViaSsh($command_line);
+        $attempt = 0;
+        $max_attempts = $retries + 1;
 
-        $this->log()->notice('Command: {site}.{env} -- {command} [Exit: {exit}]', [
-            'site'    => $this->site->getName(),
-            'env'     => $this->environment->id,
-            'command' => $command_summary,
-            'exit'    => $ssh_data['exit_code'],
-        ]);
+        do {
+            $ssh_data = $this->sendCommandViaSsh($command_line);
 
-        if ($ssh_data['exit_code'] != 0) {
-            throw new TerminusProcessException($ssh_data['output'], [], $ssh_data['exit_code']);
-        }
+            $this->log()->notice(
+                'Command: {site}.{env} -- {command} [Exit: {exit}] (Attempt {attempt}/{max_attempts})',
+                [
+                    'site' => $this->site->getName(),
+                    'env' => $this->environment->id,
+                    'command' => $command_summary,
+                    'exit' => $ssh_data['exit_code'],
+                    'attempt' => $attempt + 1,
+                    'max_attempts' => $max_attempts,
+                ]
+            );
+
+            if ($ssh_data['exit_code'] == 0) {
+                return;
+            }
+
+            // Check if the failure is permanent
+            if ($this->isPermanentFailure($ssh_data['exit_code'])) {
+                $this->log()->error('Permanent failure detected. Aborting retries. Exit code: {exit_code}', [
+                    'exit_code' => $ssh_data['exit_code']
+                ]);
+                break;
+            }
+
+            $attempt++;
+        } while ($attempt < $max_attempts);
+
+        $error_message = sprintf(
+            'Command: %s.%s -- %s [Exit: %d] (All attempts failed)',
+            $this->site->getName(),
+            $this->environment->id,
+            $command_summary,
+            $ssh_data['exit_code']
+        );
+
+        $this->log()->error($error_message);
+
+        throw new TerminusProcessException($ssh_data['output'], [], $ssh_data['exit_code']);
+    }
+
+    /**
+     * Determines if a failure is permanent based on the exit code.
+     *
+     * @param int $exit_code
+     * @return bool
+     */
+    protected function isPermanentFailure(int $exit_code): bool
+    {
+        // Define the exit codes that indicate a permanent failure
+        $permanent_failure_exit_codes = [
+            2,   // Invalid arguments
+            126, // Command cannot execute
+            127, // Command not found
+            255  // Service unavailable
```

Comment:
```suggestion
        $permanent_failure_exit_codes = [
            2,   // Invalid arguments
            126, // Command cannot execute
            127 // Command not found
```

I think we want to remove 255, because I've been able to reproduce that with what presumably should be an out-of-memory error:
```
terminus drush <SITE.ENV> php-eval '$a="asdf"; while(true) { $a .= $a; }'
```